File: example/example_machine.py
from orca import Machine, Neuron, Input


def custom_machine(name):
    pass


# these starting_* Neurons will form the base Inputs of the Machine
# we specify initial values since we dont know how the Machine will
# be hooked together.
starting_1 = Neuron(
    "starting_1",
    "u32",
    "example_machine.rs/starting_1",
    Input("data_1", "u32", constant=True, initial_value=10),
    Input("data_2", "u32", initial_value=1),
)

starting_2 = Neuron(
    "starting_2",
    "u32",
    "example_machine.rs/starting_2",
    Input("data_1", "u32", constant=True, initial_value=5),
    Input("data_2", "u32", initial_value=1),
    # Neurons can have small instances of Machines inside that they can use to
    # do calculation. TODO How do I wire up with it?
    # machines=[
    #     custom_machine(
    #         "some_machine_instance",
    #     )
    # ],
)


# Typical for people to inherit and write custom behavior for
# neurons.
class MyNeuron(Neuron):
    def __init__(self, start_value):
        Neuron.__init__(
            self,
            f"intermediate_{start_value}",
            "u32",
            "example_machine.rs/intermediate",
            Input("starting_1", "u32", query=".starting_1"),
            Input("starting_2", "u32", query=".starting_2"),
            Input("internal", "u32", initial_value=start_value),
            Input("feedback_1", "u32", query=".output_1", feedback=True),
        )


many_neurons = [MyNeuron(i) for i in range(0, 10)]


output_1 = Neuron(
    "output_1",
    "u32",
    "example_machine.rs/output_1",
    # TODO actually make this query match
    Input("many_into_one", "u32", many=10, query=".intermediate_*"),
)

# Machines can have many outputs, the Machine fires altogether.
example_machine = Machine(
    "example_machine",
    starting_1,
    starting_2,
    output_1,
    *many_neurons,
)

File: example/example_machine.rs
pub fn starting_1(data_2: u32) -> u32 {
    const data_1: u32 = 10;

    return data_2 * 2 + data_1;
}

pub fn starting_2(data_2: u32) -> u32 {
    const data_1: u32 = 5;

    return data_2 * 4 + data_1;
}

pub intermediate(starting_1: u32, starting_2: u32, internal: u32, feedback_1: u32) -> u32 {
    let new_internal = 3 * starting_1 - internal;

    set!("internal", new_internal);

    return starting_1 + starting_2 - new_internal + feedback_1;
}

pub fn output_1(many_into_one: [10; u32]) -> u32 {
    return many_into_one.sum();
}